staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go

```diff
@@ -0,0 +1,310 @@
+		intreePV *v1.PersistentVolume
+		csiPV    *v1.PersistentVolume
+		expErr   bool
+	}{
+		{
+			name:     "expect error when in-tree vsphere PV is nil",
+			intreePV: &v1.PersistentVolume{},
+			csiPV:    nil,
+			expErr:   true,
+		},
+		{
+			name: "translate valid vSphere in-tree PV to vSphere CSI PV",
+			intreePV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+							VolumePath:        "[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk",
+							FSType:            "ext4",
+							StoragePolicyName: "vSAN Default Storage Policy",
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+				},
+			},
+			csiPV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						CSI: &v1.CSIPersistentVolumeSource{
+							Driver:       VSphereDriverName,
+							VolumeHandle: "[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk",
+							FSType:       "ext4",
+							VolumeAttributes: map[string]string{
+								paramStoragePolicyName: "vSAN Default Storage Policy",
+							},
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+				},
+			},
+			expErr: false,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Logf("Testing %v", tc.name)
+		got, err := translator.TranslateInTreePVToCSI(tc.intreePV)
+		if err != nil && !tc.expErr {
+			t.Errorf("Did not expect error but got: %v", err)
+		}
+
+		if err == nil && tc.expErr {
+			t.Errorf("Expected error, but did not get one.")
+		}
+
+		if !reflect.DeepEqual(got, tc.csiPV) {
+			t.Errorf("Got PV: %v, expected :%v", got, tc.csiPV)
+		}
+	}
+}
+
+func TestTranslatevSphereInTreeInlineVolumeToCSI(t *testing.T) {
+	translator := NewvSphereCSITranslator()
+	cases := []struct {
+		name         string
+		inlinevolume *v1.Volume
+		csiPV        *v1.PersistentVolume
+		expErr       bool
+	}{
+		{
+			name:         "expect error when inline vsphere volume is nil",
+			inlinevolume: &v1.Volume{},
+			csiPV:        nil,
+			expErr:       true,
+		},
+		{
+			name: "translate valid in-tree vsphere volume to vSphere CSI PV",
+			inlinevolume: &v1.Volume{
+				Name: "inlinevolume",
+				VolumeSource: v1.VolumeSource{
+					VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+						VolumePath: "[vsanDatastore] volume/inlinevolume.vmdk",
+						FSType:     "ext4",
+					},
+				}},
+			csiPV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: fmt.Sprintf("%s-%s", VSphereDriverName, "[vsanDatastore] volume/inlinevolume.vmdk"),
```

Comment:
Did you confirm that spaces and square brackets are valid characters for resource names?

EDIT: I guess it's actually OK because the translated object never goes to the API server. Just seemed a bit weird.

Reply:
since PV is not created using this name. it is ok.